funcs.php

```diff
@@ -0,0 +1,111 @@
+<?php
+
+function branches(
+    string $defaultBranch,
+    string $minimumCmsMajor,
+    // The following params are purely for unit testing, for the actual github action it will read json files instead
+    string $composerJson = '',
+    string $branchesJson = '',
+    string $tagsJson = ''
+) {
+    if (!is_numeric($defaultBranch)) {
+        throw new Exception('Default branch must be a number');
+    }
+    if (!ctype_digit($minimumCmsMajor)) {
+        throw new Exception('Minimum CMS major must be an integer');
+    }
+
+    // work out default major
+    preg_match('#^([0-9]+)+\.?[0-9]*$#', $defaultBranch, $matches);
+    $defaultMajor = $matches[1];
+    
+    // read __composer.json of the current (default) branch
+    $contents = $composerJson ?: file_get_contents('__composer.json');
+    $json = json_decode($contents);
+    if (is_null($json)) {
+        throw new Exception('Could not parse __composer.json');
+    }
+    $defaultCmsMajor = '';
+    $version = preg_replace('#[^0-9\.]#', '', $json->require->{'silverstripe/framework'} ?? '');
+    if (preg_match('#^([0-9]+)+\.?[0-9]*$#', $version, $matches)) {
+        $defaultCmsMajor = $matches[1];
+    } else {
+        $phpVersion = $json->require->{'php'} ?? '';
+        if (substr($phpVersion,0, 4) === '^7.4') {
+            $defaultCmsMajor = 4;
+        } elseif (substr($phpVersion,0, 4) === '^8.1') {
+            $defaultCmsMajor = 5;
+        }
+    }
+    if ($defaultCmsMajor === '') {
+        throw new Exception('Could not work out what the default CMS major version this module uses');
+    }
```

Comment:
We could, though it seems like overkill since everything just requires framework - i.e. we need to copy in this https://github.com/silverstripe/gha-generate-matrix/blob/1/consts.php#L87

CI need to be more robust because it could be used by the community for use on any module. However gha-merge-ups is only for use on supported modules that we look after ... and everything just requires framework

Reply:
I'm not saying copy, I'm saying reuse as in have it in some central place where we can just update logic or mapping once and it'll be reflected in both places. My main problem with this implementation is its reliance on PHP versions, which means we have _yet another_ thing we have to remember to update (or wait for it to break) next time we do a major release.

If there's no clean way to put that logic or mapping in a central location, then I'll accept this implementation.

Reply:
It's not necessary IMO and adds a lot of complexity to this since we'd new too create another github-action just for doing determining the version, and then updating this script to use it, and then also updating gha-generate-matrix to also use it

Reply:
Fair enough. I think it reduces complexity by having only one set of variables and logic for the same functionality that we need to maintain - but I won't press the point. This can stay as is.